internal/notification/slack/slack.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/rs/zerolog/log"
+	"github.com/slack-go/slack"
+
+	"github.com/saucelabs/saucectl/internal/config"
+	"github.com/saucelabs/saucectl/internal/credentials"
+)
+
+// Reporter represents reporter for slack
+type Reporter struct {
+	Channels       []string
+	TestResults    []report.TestResult
+	Framework      string
+	Metadata       config.Metadata
+	TestEnv        string
+	RenderedResult string
+	Config         config.Notifications
+	Service        Service
+	lock           sync.Mutex
+}
+
+// Add adds the TestResult to the reporter. TestResults added this way can then be rendered out by calling Render().
+func (r *Reporter) Add(t report.TestResult) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.TestResults = append(r.TestResults, t)
+}
+
+// Render renders the test results and sends message to the Slack.
+func (r *Reporter) Render() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	tables := [][]string{}
+	longestName := 0
+	for _, ts := range r.TestResults {
+		if longestName < len(ts.Name) {
+			longestName = len(ts.Name)
+		}
+	}
+
+	passed := true
+	for _, ts := range r.TestResults {
+		tables = append(tables, []string{statusText(ts.Passed), regenerateName(ts.Name, r.getJobURL(ts.Name, ts.URL), longestName),
+			ts.Platform, ts.DeviceName, ts.Browser, ts.Duration.Truncate(1 * time.Second).String()})
+		if !ts.Passed {
+			passed = false
+		}
+	}
+	var res string
+	for _, t := range tables {
+		res = fmt.Sprintf("%s\n%s", res, strings.Join(t, "\t"))
+	}
+
+	r.RenderedResult = res
+
+	r.sendMessage(passed)
+}
+
+// GetRenderedResult returns rendered result.
+func (r *Reporter) GetRenderedResult() string {
+	return r.RenderedResult
+}
+
+// Reset resets the state of the reporter (e.g. remove any previously reported TestResults).
+func (r *Reporter) Reset() {}
+
+// ArtifactRequirements returns a list of artifact types that this reporter requires to create a proper report.
+func (r *Reporter) ArtifactRequirements() []report.ArtifactType {
+	return nil
+}
+
+// sendMessage sends notification message.
+func (r *Reporter) sendMessage(passed bool) {
+	if !r.shouldSendNotification(passed) {
+		return
+	}
+
+	token, err := r.Service.GetSlackToken(context.Background())
+	if err != nil {
+		log.Err(err).Msg("Failed to get slack token")
+		return
+	}
+
+	api := slack.New(token)
+
+	for _, c := range r.Channels {
+		_, timestamp, err := api.PostMessage(
+			c,
+			slack.MsgOptionText("saucectl test result", false),
+			slack.MsgOptionAttachments(r.creatAttachment(passed)),
+			slack.MsgOptionAsUser(true),
+		)
+
+		if err != nil {
+			log.Error().Msgf("Failed to send message to slack, err: %s", err.Error())
+			continue
+		}
+
+		timestampArr := strings.Split(timestamp, ".")
+		i, err := strconv.ParseInt(timestampArr[0], 10, 64)
+		if err != nil {
+			log.Info().Msgf("Couldn't parse slack timestamp, err: %s", err)
+		} else {
+			i = time.Now().Unix()
+		}
+
+		log.Info().Msgf("Message successfully sent to slack channel %s at %s", c, time.Unix(i, 0))
```

Comment:
See my comment above ^ 
```suggestion
		log.Info().Msgf("Message successfully sent to slack channel %s", c)
```

Reply:
done